myia/ir.py

```diff
@@ -1,7 +1,142 @@
-"""Intermediate representation definition."""
+"""Intermediate representation definition.
+
+Myia's main intermediate representation (IR) is a graph-based version of ANF.
+Each function definition (lambda) is defined as a graph, consisting of a series
+of function applications.
+
+A function can be applied to a node from another funtion's graph; this
+implicitly creates a nested function. Functions are first-class objects, so
+returning a nested function creates a closure.
+
+"""
+from enum import Enum, auto
+from typing import List, Union, Dict
+
+LiteralType = Union[str, int, float, 'PrimitiveOperation', 'Graph', None]
+
+
+class Graph:
+    """A function graph.
+
+    Attributes:
+        output: The `Node` whose value will be returned by this function.
+
+    """
+
+    def __init__(self, output: 'Node') -> None:
```

Comment:
Two things I'd missed the first time around:

* A `Graph` should have a list of parameters.
* You have to be able to create a `Graph` before you know its output, so that recursive applications can have a handle to it (alternatively a `Constant` node could be created, with its graph pending).

Reply:
Makes sense. I initially figured that a graph could just find its parameters by walking from the output to the inputs, but I guess that doesn't work if there are unused parameters.

Reply:
That too, but the main problem with walking from the output to the inputs is that it yields them in an arbitrary order.